File: backend/app/models/project.py
from datetime import datetime, timezone
from decimal import Decimal

from datetime import date

from sqlalchemy import Date, DateTime, Integer, Numeric, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base


class Project(Base):

    __tablename__ = "projects"

    id: Mapped[int] = mapped_column(
        Integer,
        primary_key=True,
        autoincrement=True,
    )

    name: Mapped[str] = mapped_column(
        String(250),
        nullable=False,
    )

    state: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
    )

    district: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
    )

    proposed_area_ha: Mapped[Decimal] = mapped_column(
        Numeric(14, 4),
        nullable=False,
    )

    status: Mapped[str] = mapped_column(
        String(30),
        nullable=False,
        default="DRAFT",
    )

    owner_username: Mapped[str | None] = mapped_column(String(80), index=True)
    agency: Mapped[str | None] = mapped_column(String(180))
    sector: Mapped[str | None] = mapped_column(String(80))
    description: Mapped[str | None] = mapped_column(Text)
    target_date: Mapped[date | None] = mapped_column(Date)

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        nullable=False,
    )

File: frontend/src/types/project.ts
export type Project = {
  id: number
  name: string
  state: string
  district: string
  proposed_area_ha: string
  status: string
  agency: string | null
  sector: string | null
  description: string | null
  target_date: string | null
  owner_username: string | null
  created_at: string
}

export type ProjectInput = {
  name: string
  state: string
  district: string
  proposed_area_ha: number
  agency?: string | null
  sector?: string | null
  description?: string | null
  target_date?: string | null
}
